proposals/2676-version-dataset-schemas-separately.md

```diff
@@ -0,0 +1,170 @@
+
+Taking an oversimplified example, a dataset with a stable 20-column schema that is written to by a job that runs every 10 minutes for 30 days will result in 864,000 rows in `dataset_versions_field_mapping` when arguably there are only 20 meaningful rows of data. This is regardless of how many of those columns are actually touched by the job.
+
+A dataset must still be versioned each time it occurs as an output, since it may contain facets that relate to the state of its structure and/or data at that moment in time beyond just the schema - including column-level lineage. But the schema itself doesn't need to be re-versioned unless it has actually changed.
+
+## Proposal
+
+(This is not intended to alter any user-facing behaviour, only to reduce the volume and duplication at the database level.)
+
+The concept of a "dataset schema version" is added to the model. Its content is the fields belonging to the dataset schema, in the form of a mappings to dataset fields. We'll refer to this as a "schema version" for brevity.
+
+The first time a dataset is encountered as a job output (or, the next time after deploying this change), a schema version is created, and associated with the new dataset version. The schema version is the source of truth for fields for that dataset version when queried.
+
+Each time that dataset is encountered as a job output thereafter, the schema is compared against the current schema version. If it is considered equal, a new schema version isn't created, and the new dataset version references the current schema version.
+
+With the oversimplified example above, this would then result in:
+
+- 1 row in `dataset_schema_versions`
+- 20 rows in `dataset_schema_versions_field_mapping`
+
+### Equality
+
+Two schemas will be considered equal if they contain the same fields, as keyed by `name` and `type` - as such, a changed `description` is not considered a schema change. Order is also not significant - we would order the fields alphabetically before doing the check.
+
+### Input datasets
+
+When Marquez encounters a dataset as an input in an event, and the dataset already exists, it will not re-version it (as it would for an output), and thus any schema differences are merged onto the current dataset version. When we introduce dataset schema versions, we do _not_ want to modify the existing schema version in this case, as its contents should be immutable. Instead, we would create a new schema version with the amended schema, and update the dataset version to reference that schema version.
+
+## Implementation
+
+### Schema changes
+
+- New table `dataset_schema_versions` added with columns:
+  - `created_at` (timestamp)
+  - `schema_hash` (content hash for equality checking)
+  - `dataset_uuid` (fk to `datasets`)
+  - `uuid` (pk)
```

Comment:
To avoid the column `schema_hash`, can we generate the "hash" as a `UUID` as we do, for example, job [versions](https://github.com/MarquezProject/marquez/blob/main/api/src/main/java/marquez/common/Utils.java) using `UUID.nameUUIDFromBytes(bytes)`?

Reply:
Oops, I didn't finish my thought. If we hash to a `UUID`, we can use that value as the `uuid` for the schema version in place of `schema_hash` (that we'd no longer need).

Reply:
Ah okay, I'll take a look at this an update it!

Reply:
Updated now.

Reply:
Thanks!